File: src/judging/codec/encode.ts
/**
 * Canonical encoder: {@link CanonicalValue} to canonical JSON text.
 *
 * The encoder is the authoritative definition of the `tagged-jsonl-v1` wire
 * form. It guarantees determinism: object keys are emitted in UTF-8 lexical
 * order, set/frozenset items and dict entries are sorted (and de-duplicated) by
 * canonical encoded bytes, and every leaf is validated against its normative
 * grammar before emission. Equal values therefore always produce byte-identical
 * output, which is what makes differential testing and byte-equality policies
 * meaningful.
 *
 * Encoding is strict: an invalid in-memory value (a non-canonical leaf, a
 * reference cycle, duplicate set/dict members, or an exceeded limit) throws
 * {@link CodecEncodeError} rather than emitting malformed bytes.
 */

import type {
  CanonicalValue,
  DictEntry,
  NumericLeaf,
} from "../value/model.js";
import { Budget, LimitExceededError } from "./limits.js";
import { CodecEncodeError } from "./errors.js";
import {
  decodeBase64Url,
  encodeBase64Url,
  isCanonicalFloat,
  isCanonicalRelativePath,
  isValidDate,
  isValidDatetimeText,
  isValidDecimalLiteral,
  isValidOffsetMinutes,
  isValidTimeOfDay,
  isValidUuid,
} from "./leaf-grammar.js";
import { compareUtf8, hasLoneSurrogate } from "./utf8.js";

/**
 * Largest integer magnitude that may be emitted as a bare JSON number.
 *
 * This is `2^53 - 1`, the edge of the IEEE-754 exact-integer range. Integers
 * outside `[-MAX_SAFE_INT, MAX_SAFE_INT]` are emitted as decimal strings so no
 * precision is lost when a lax reader parses the JSON with a native number.
 */
const MAX_SAFE_INT = 9_007_199_254_740_991n;

/**
 * Encode a canonical value to its canonical JSON text.
 *
 * @param value - The value to encode.
 * @param budget - Depth/node accountant for this encode traversal.
 * @returns The canonical JSON text (no trailing newline).
 * @throws {CodecEncodeError} If the value is not canonically encodable.
 */
export function encodeValue(value: CanonicalValue, budget: Budget): string {
  try {
    return encodeNode(value, budget, new Set<object>(), "$");
  } catch (err) {
    if (err instanceof LimitExceededError) {
      // Present limit violations uniformly as an encode failure.
      throw new CodecEncodeError(err.message);
    }
    throw err;
  }
}

/**
 * Encode a value using a throwaway budget.
 *
 * Used to obtain the canonical bytes of a sub-value for ordering/uniqueness
 * comparisons (sets, dict keys) without perturbing an in-progress budget.
 *
 * @param value - The value to encode.
 * @returns The canonical JSON text.
 */
export function canonicalStringOf(value: CanonicalValue): string {
  return encodeValue(value, new Budget());
}

/** Recursive worker; `seen` holds ancestor containers to detect cycles. */
function encodeNode(
  value: CanonicalValue,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  budget.countNode();
  switch (value.tag) {
    case "null":
      return '{"tag":"null"}';
    case "bool":
      return obj([["tag", '"bool"'], ["value", value.value ? "true" : "false"]]);
    case "int":
      return obj([["tag", '"int"'], ["value", encodeInt(value.value)]]);
    case "float":
      return encodeFloat(value.value, value.negativeZero, path);
    case "decimal":
      if (!isValidDecimalLiteral(value.value)) {
        throw new CodecEncodeError("invalid decimal literal", path);
      }
      return obj([["tag", '"decimal"'], ["value", jsonString(value.value)]]);
    case "complex":
      return obj([
        ["imag", encodeLeaf(value.imag, budget, seen, `${path}.imag`)],
        ["real", encodeLeaf(value.real, budget, seen, `${path}.real`)],
        ["tag", '"complex"'],
      ]);
    case "str":
      return obj([["tag", '"str"'], ["value", encodeStr(value.value, path)]]);
    case "list":
    case "tuple":
      return encodeSequence(value.tag, value.items, budget, seen, path);
    case "set":
    case "frozenset":
      return encodeSet(value.tag, value.items, budget, seen, path);
    case "dict":
      return encodeDict(value.entries, budget, seen, path);
    case "bytes":
      return encodeBytes(value.value, path);
    case "date":
      if (!isValidDate(value.value)) {
        throw new CodecEncodeError("invalid date", path);
      }
      return obj([["tag", '"date"'], ["value", jsonString(value.value)]]);
    case "time":
      return encodeTime(value.value, value.offsetMinutes, value.fold, path);
    case "datetime":
      return encodeDatetime(value.value, value.offsetMinutes, value.fold, path);
    case "uuid":
      if (!isValidUuid(value.value)) {
        throw new CodecEncodeError("invalid uuid", path);
      }
      return obj([["tag", '"uuid"'], ["value", jsonString(value.value)]]);
    case "path":
      if (!isCanonicalRelativePath(value.value, value.flavor)) {
        throw new CodecEncodeError("invalid or absolute path", path);
      }
      return obj([
        ["flavor", jsonString(value.flavor)],
        ["tag", '"path"'],
        ["value", jsonString(value.value)],
      ]);
    case "enum":
      return withContainer(value, seen, path, () =>
        obj([
          ["member", encodeStr(value.member, `${path}.member`)],
          ["tag", '"enum"'],
          ["type", encodeStr(value.type, `${path}.type`)],
          ["value", encodeNode(value.value, budget, seen, `${path}.value`)],
        ]),
      );
    case "record":
      return encodeRecord(value, budget, seen, path);
    case "exception":
      return encodeException(value, budget, seen, path);
    case "ListNode":
      return encodeListNode(value, budget, seen, path);
    case "TreeNode":
      return encodeTreeNode(value, budget, seen, path);
    case "ClassTrace":
      return encodeClassTrace(value, budget, seen, path);
  }
}

/** Encode a required numeric leaf, rejecting any non-numeric tag. */
function encodeLeaf(
  leaf: NumericLeaf,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  if (leaf.tag !== "int" && leaf.tag !== "float" && leaf.tag !== "decimal") {
    throw new CodecEncodeError("complex parts must be numeric leaves", path);
  }
  return encodeNode(leaf, budget, seen, path);
}

/** Emit a bigint as a bare JSON number when safe, else as a decimal string. */
function encodeInt(value: bigint): string {
  if (value >= -MAX_SAFE_INT && value <= MAX_SAFE_INT) {
    return value.toString();
  }
  return jsonString(value.toString());
}

/** Validate and emit a canonical float object. */
function encodeFloat(value: string, negativeZero: boolean, path: string): string {
  if (!isCanonicalFloat(value, negativeZero)) {
    throw new CodecEncodeError("non-canonical or non-finite float", path);
  }
  return obj([
    ["negativeZero", negativeZero ? "true" : "false"],
    ["tag", '"float"'],
    ["value", jsonString(value)],
  ]);
}

/** Validate and emit a canonical str value payload (rejecting lone surrogates). */
function encodeStr(value: string, path: string): string {
  if (hasLoneSurrogate(value)) {
    throw new CodecEncodeError("string contains a lone surrogate", path);
  }
  return jsonString(value);
}

/** Encode an ordered `list`/`tuple` whose element order is significant. */
function encodeSequence(
  tag: "list" | "tuple",
  items: readonly CanonicalValue[],
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer({ tag, items } as object, seen, path, () => {
    budget.enter();
    const parts = items.map((item, i) =>
      encodeNode(item, budget, seen, `${path}.items[${i}]`),
    );
    budget.leave();
    return obj([["items", `[${parts.join(",")}]`], ["tag", jsonString(tag)]]);
  });
}

/** Encode a `set`/`frozenset`: items sorted by encoded bytes and unique. */
function encodeSet(
  tag: "set" | "frozenset",
  items: readonly CanonicalValue[],
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer({ tag, items } as object, seen, path, () => {
    budget.enter();
    const encoded = items.map((item, i) =>
      encodeNode(item, budget, seen, `${path}.items[${i}]`),
    );
    budget.leave();
    encoded.sort(compareUtf8);
    for (let i = 1; i < encoded.length; i += 1) {
      if (compareUtf8(encoded[i - 1] as string, encoded[i] as string) === 0) {
        throw new CodecEncodeError(`duplicate ${tag} member`, path);
      }
    }
    return obj([["items", `[${encoded.join(",")}]`], ["tag", jsonString(tag)]]);
  });
}

/** Encode a `dict`: entries sorted by encoded key bytes, keys unique. */
function encodeDict(
  entries: readonly DictEntry[],
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer({ tag: "dict", entries } as object, seen, path, () => {
    budget.enter();
    const encoded = entries.map((entry, i) => {
      const key = encodeNode(entry.key, budget, seen, `${path}.entries[${i}].key`);
      const val = encodeNode(
        entry.value,
        budget,
        seen,
        `${path}.entries[${i}].value`,
      );
      return { key, entry: obj([["key", key], ["value", val]]) };
    });
    budget.leave();
    encoded.sort((a, b) => compareUtf8(a.key, b.key));
    for (let i = 1; i < encoded.length; i += 1) {
      const prev = encoded[i - 1];
      const cur = encoded[i];
      if (prev !== undefined && cur !== undefined && compareUtf8(prev.key, cur.key) === 0) {
        throw new CodecEncodeError("duplicate dict key", path);
      }
    }
    return obj([
      ["entries", `[${encoded.map((e) => e.entry).join(",")}]`],
      ["tag", '"dict"'],
    ]);
  });
}

/** Validate and emit a canonical bytes object. */
function encodeBytes(value: string, path: string): string {
  // Re-encoding what we decode guarantees canonical unpadded base64url output.
  return obj([
    ["encoding", '"base64url"'],
    ["tag", '"bytes"'],
    ["value", jsonString(normalizeBase64Url(value, path))],
  ]);
}

/** Ensure a bytes payload is canonical base64url; return it verbatim. */
function normalizeBase64Url(value: string, path: string): string {
  // A canonical value already holds canonical text; recompute defensively so a
  // corrupt in-memory payload cannot slip onto the wire.
  const bytes = decodeBase64Url(value);
  if (bytes === null || encodeBase64Url(bytes) !== value) {
    throw new CodecEncodeError("invalid base64url", path);
  }
  return value;
}

/** Validate and emit a canonical time object. */
function encodeTime(
  value: string,
  offsetMinutes: number | null,
  fold: 0 | 1,
  path: string,
): string {
  if (!isValidTimeOfDay(value)) {
    throw new CodecEncodeError("invalid time", path);
  }
  if (offsetMinutes !== null && !isValidOffsetMinutes(offsetMinutes)) {
    throw new CodecEncodeError("offsetMinutes out of range", path);
  }
  return obj([
    ["fold", String(fold)],
    ["offsetMinutes", offsetMinutes === null ? "null" : String(offsetMinutes)],
    ["tag", '"time"'],
    ["value", jsonString(value)],
  ]);
}

/** Validate and emit a canonical datetime object. */
function encodeDatetime(
  value: string,
  offsetMinutes: number,
  fold: 0 | 1,
  path: string,
): string {
  if (!isValidDatetimeText(value)) {
    throw new CodecEncodeError("invalid datetime", path);
  }
  if (!isValidOffsetMinutes(offsetMinutes)) {
    throw new CodecEncodeError("offsetMinutes out of range", path);
  }
  return obj([
    ["fold", String(fold)],
    ["offsetMinutes", String(offsetMinutes)],
    ["tag", '"datetime"'],
    ["value", jsonString(value)],
  ]);
}

/** Encode a `record` with declaration-ordered, uniquely named fields. */
function encodeRecord(
  value: Extract<CanonicalValue, { tag: "record" }>,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer(value, seen, path, () => {
    budget.enter();
    const seenNames = new Set<string>();
    const fields = value.fields.map((field, i) => {
      if (seenNames.has(field.name)) {
        throw new CodecEncodeError("duplicate record field name", path);
      }
      seenNames.add(field.name);
      return obj([
        ["name", encodeStr(field.name, `${path}.fields[${i}].name`)],
        ["value", encodeNode(field.value, budget, seen, `${path}.fields[${i}].value`)],
      ]);
    });
    budget.leave();
    return obj([
      ["fields", `[${fields.join(",")}]`],
      ["name", encodeStr(value.name, `${path}.name`)],
      ["tag", '"record"'],
      ["type", jsonString(value.type)],
    ]);
  });
}

/** Encode an `exception` value: type, message, and optional tagged details. */
function encodeException(
  value: Extract<CanonicalValue, { tag: "exception" }>,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer(value, seen, path, () => {
    const details =
      value.details === null
        ? "null"
        : encodeNode(value.details, budget, seen, `${path}.details`);
    return obj([
      ["details", details],
      ["message", encodeStr(value.message, `${path}.message`)],
      ["tag", '"exception"'],
      ["type", encodeStr(value.type, `${path}.type`)],
    ]);
  });
}

/** Encode a `ListNode` adapter value; `cycleIndex` must point within `values`. */
function encodeListNode(
  value: Extract<CanonicalValue, { tag: "ListNode" }>,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer(value, seen, path, () => {
    budget.enter();
    const values = value.values.map((v, i) =>
      encodeNode(v, budget, seen, `${path}.values[${i}]`),
    );
    budget.leave();
    if (value.cycleIndex !== null) {
      if (
        !Number.isInteger(value.cycleIndex) ||
        value.cycleIndex < 0 ||
        value.cycleIndex >= value.values.length
      ) {
        throw new CodecEncodeError("cycleIndex out of range", path);
      }
    }
    return obj([
      ["cycleIndex", value.cycleIndex === null ? "null" : String(value.cycleIndex)],
      ["tag", '"ListNode"'],
      ["values", `[${values.join(",")}]`],
    ]);
  });
}

/** Encode a `TreeNode` adapter value, stripping trailing nulls canonically. */
function encodeTreeNode(
  value: Extract<CanonicalValue, { tag: "TreeNode" }>,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer(value, seen, path, () => {
    // Strip trailing null slots so the canonical form is unique.
    const slots = value.values.slice();
    while (slots.length > 0 && slots[slots.length - 1] === null) {
      slots.pop();
    }
    // A null parent may not have a non-null descendant (unreachable node).
    for (let i = 0; i < slots.length; i += 1) {
      if (slots[i] === null) {
        const left = 2 * i + 1;
        const right = 2 * i + 2;
        if (
          (left < slots.length && slots[left] !== null) ||
          (right < slots.length && slots[right] !== null)
        ) {
          throw new CodecEncodeError("unreachable TreeNode node", path);
        }
      }
    }
    budget.enter();
    const parts = slots.map((slot, i) =>
      slot === null ? "null" : encodeNode(slot, budget, seen, `${path}.values[${i}]`),
    );
    budget.leave();
    return obj([["tag", '"TreeNode"'], ["values", `[${parts.join(",")}]`]]);
  });
}

/** Encode a `ClassTrace` adapter value: constructor args then operations. */
function encodeClassTrace(
  value: Extract<CanonicalValue, { tag: "ClassTrace" }>,
  budget: Budget,
  seen: Set<object>,
  path: string,
): string {
  return withContainer(value, seen, path, () => {
    budget.enter();
    const ctor = value.constructor.map((v, i) =>
      encodeNode(v, budget, seen, `${path}.constructor[${i}]`),
    );
    const ops = value.operations.map((op, i) => {
      const args = op.args.map((a, j) =>
        encodeNode(a, budget, seen, `${path}.operations[${i}].args[${j}]`),
      );
      const pairs: [string, string][] = [
        ["args", `[${args.join(",")}]`],
        ["method", encodeStr(op.method, `${path}.operations[${i}].method`)],
      ];
      if (op.expected !== undefined) {
        pairs.push([
          "expected",
          encodeNode(op.expected, budget, seen, `${path}.operations[${i}].expected`),
        ]);
      }
      return obj(pairs);
    });
    budget.leave();
    return obj([
      ["className", encodeStr(value.className, `${path}.className`)],
      ["constructor", `[${ctor.join(",")}]`],
      ["operations", `[${ops.join(",")}]`],
      ["tag", '"ClassTrace"'],
    ]);
  });
}

/** Run `body` with `node` pushed on the cycle-detection path. */
function withContainer(
  node: object,
  seen: Set<object>,
  path: string,
  body: () => string,
): string {
  if (seen.has(node)) {
    throw new CodecEncodeError("reference cycle detected", path);
  }
  seen.add(node);
  try {
    return body();
  } finally {
    seen.delete(node);
  }
}

/**
 * Assemble a JSON object from `[key, encodedValue]` pairs with keys emitted in
 * UTF-8 lexical order. The value strings are already-encoded JSON fragments.
 */
function obj(pairs: readonly (readonly [string, string])[]): string {
  const sorted = pairs
    .slice()
    .sort((a, b) => compareUtf8(a[0], b[0]));
  return `{${sorted.map(([k, v]) => `${jsonString(k)}:${v}`).join(",")}}`;
}

/** Encode a JSON string leaf deterministically via the platform serializer. */
function jsonString(value: string): string {
  return JSON.stringify(value);
}

File: src/execution/runtimes/python/harness/canonical/encode.py
"""Native Python objects -> canonical ``tagged-jsonl-v1`` JSON text.

Mirror of ``src/judging/codec/encode.ts``. :func:`encode_value` is the
authoritative Python side of the wire form: object keys are emitted in UTF-8
lexical order, ``set``/``frozenset`` items and ``dict`` entries are sorted (and
de-duplicated) by canonical encoded bytes, and every leaf is validated against
its normative grammar before emission. Equal values therefore serialize to
byte-identical output, matching the TypeScript encoder so byte-equality judging
holds across the process boundary.

Encoding is strict: a non-representable value (NaN/Infinity float, non-canonical
leaf, reference cycle, duplicate set/dict member, or exceeded limit) raises
:class:`CodecEncodeError`.
"""

from __future__ import annotations

import dataclasses
import math
from datetime import date, datetime, time
from decimal import Decimal
from enum import Enum
from pathlib import PurePath, PureWindowsPath
from typing import Any
from uuid import UUID

from . import grammar
from .adapters import ListNode, TreeNode, list_to_canonical, tree_to_canonical
from .grammar import Budget, LimitExceededError
from .jsonio import has_lone_surrogate, json_string


class CodecEncodeError(Exception):
    """Raised when a native value cannot be canonically encoded."""

    def __init__(self, message: str, path: str = "$") -> None:
        super().__init__(message)
        self.path = path


def encode_value(value: Any, budget: Budget | None = None) -> str:
    """Encode a native value to its canonical JSON text (no trailing newline).

    :param value: The native Python value to encode.
    :param budget: Optional depth/node accountant; a fresh one is used if omitted.
    :raises CodecEncodeError: If the value is not canonically encodable.
    """
    accountant = budget if budget is not None else Budget()
    try:
        return _encode_node(value, accountant, set(), "$")
    except LimitExceededError as err:
        raise CodecEncodeError(str(err)) from err


def _obj(pairs: list[tuple[str, str]]) -> str:
    """Assemble a JSON object from ``(key, encoded_value)`` pairs.

    Keys are emitted in UTF-8 byte order (Python ``bytes`` comparison is
    lexicographic, matching ``compareUtf8``); the values are already-encoded JSON
    fragments.
    """
    pairs.sort(key=lambda pair: pair[0].encode("utf-8"))
    return "{" + ",".join(f"{json_string(k)}:{v}" for k, v in pairs) + "}"


def _encode_node(value: Any, budget: Budget, seen: set[int], path: str) -> str:
    budget.count_node()

    if value is None:
        return '{"tag":"null"}'
    if isinstance(value, bool):
        return _obj([("tag", '"bool"'), ("value", "true" if value else "false")])
    if isinstance(value, Enum):
        return _encode_enum(value, budget, seen, path)
    if isinstance(value, int):
        return _obj([("tag", '"int"'), ("value", _encode_int(value))])
    if isinstance(value, float):
        return _encode_float(value, path)
    if isinstance(value, Decimal):
        return _encode_decimal(value, path)
    if isinstance(value, complex):
        return _obj(
            [
                ("imag", _encode_float(value.imag, f"{path}.imag")),
                ("real", _encode_float(value.real, f"{path}.real")),
                ("tag", '"complex"'),
            ]
        )
    if isinstance(value, str):
        return _obj([("tag", '"str"'), ("value", _encode_str(value, path))])
    if isinstance(value, (bytes, bytearray)):
        return _encode_bytes(bytes(value), path)
    if isinstance(value, ListNode):
        return _encode_list_node(value, budget, seen, path)
    if isinstance(value, TreeNode):
        return _encode_tree_node(value, budget, seen, path)
    if isinstance(value, datetime):
        return _encode_datetime(value, path)
    if isinstance(value, date):
        return _encode_date(value, path)
    if isinstance(value, time):
        return _encode_time(value, path)
    if isinstance(value, UUID):
        return _encode_uuid(value, path)
    if isinstance(value, PurePath):
        return _encode_path(value, path)
    if _is_namedtuple(value):
        return _encode_record(value, "namedtuple", type(value).__name__, list(value._fields), list(value), budget, seen, path)
    if isinstance(value, tuple):
        return _encode_sequence("tuple", list(value), budget, seen, path)
    if isinstance(value, list):
        return _encode_sequence("list", value, budget, seen, path)
    if isinstance(value, frozenset):
        return _encode_set("frozenset", value, budget, seen, path)
    if isinstance(value, set):
        return _encode_set("set", value, budget, seen, path)
    if dataclasses.is_dataclass(value) and not isinstance(value, type):
        fields = dataclasses.fields(value)
        names = [f.name for f in fields]
        values = [getattr(value, f.name) for f in fields]
        return _encode_record(value, "dataclass", type(value).__name__, names, values, budget, seen, path)
    if isinstance(value, dict):
        return _encode_dict(value, budget, seen, path)
    if isinstance(value, BaseException):
        return _encode_exception(value, budget, seen, path)
    raise CodecEncodeError(f"unsupported value of type {type(value).__name__}", path)


def _is_namedtuple(value: Any) -> bool:
    """Whether ``value`` is a namedtuple instance (a tuple carrying ``_fields``)."""
    return isinstance(value, tuple) and hasattr(value, "_fields")


def _encode_int(value: int) -> str:
    if -grammar.MAX_SAFE_INT <= value <= grammar.MAX_SAFE_INT:
        return str(value)
    return json_string(str(value))


def _encode_float(value: float, path: str) -> str:
    if math.isnan(value) or math.isinf(value):
        raise CodecEncodeError("non-canonical or non-finite float", path)
    negative_zero = value == 0.0 and math.copysign(1.0, value) < 0.0
    text = grammar.canonical_float_text(value)
    if not grammar.is_canonical_float(text, negative_zero):
        raise CodecEncodeError("non-canonical or non-finite float", path)
    return _obj(
        [
            ("negativeZero", "true" if negative_zero else "false"),
            ("tag", '"float"'),
            ("value", json_string(text)),
        ]
    )


def _encode_decimal(value: Decimal, path: str) -> str:
    if not value.is_finite():
        raise CodecEncodeError("non-finite decimal", path)
    text = str(value)
    if not grammar.is_valid_decimal_literal(text):
        raise CodecEncodeError("invalid decimal literal", path)
    return _obj([("tag", '"decimal"'), ("value", json_string(text))])


def _encode_str(value: str, path: str) -> str:
    if has_lone_surrogate(value):
        raise CodecEncodeError("string contains a lone surrogate", path)
    return json_string(value)


def _encode_bytes(value: bytes, path: str) -> str:
    return _obj(
        [
            ("encoding", '"base64url"'),
            ("tag", '"bytes"'),
            ("value", json_string(grammar.encode_base64url(value))),
        ]
    )


def _encode_date(value: date, path: str) -> str:
    text = value.isoformat()
    if not grammar.is_valid_date(text):
        raise CodecEncodeError("invalid date", path)
    return _obj([("tag", '"date"'), ("value", json_string(text))])


def _time_text(value: time) -> str:
    """Render a ``time``'s wall-clock text ``HH:MM:SS[.ffffff]`` (no offset)."""
    text = f"{value.hour:02d}:{value.minute:02d}:{value.second:02d}"
    if value.microsecond:
        text += f".{value.microsecond:06d}"
    return text


def _offset_minutes(utc_offset: Any) -> int | None:
    """Convert a ``timedelta`` UTC offset to whole minutes, or ``None``."""
    if utc_offset is None:
        return None
    total_seconds = int(utc_offset.total_seconds())
    return total_seconds // 60


def _encode_time(value: time, path: str) -> str:
    text = _time_text(value)
    if not grammar.is_valid_time_of_day(text):
        raise CodecEncodeError("invalid time", path)
    offset = _offset_minutes(value.utcoffset())
    if offset is not None and not grammar.is_valid_offset_minutes(offset):
        raise CodecEncodeError("offsetMinutes out of range", path)
    return _obj(
        [
            ("fold", str(value.fold)),
            ("offsetMinutes", "null" if offset is None else str(offset)),
            ("tag", '"time"'),
            ("value", json_string(text)),
        ]
    )


def _encode_datetime(value: datetime, path: str) -> str:
    text = f"{value.date().isoformat()}T{_time_text(value.timetz())}"
    if not grammar.is_valid_datetime_text(text):
        raise CodecEncodeError("invalid datetime", path)
    offset = _offset_minutes(value.utcoffset())
    if offset is None:
        raise CodecEncodeError("naive datetime is not representable", path)
    if not grammar.is_valid_offset_minutes(offset):
        raise CodecEncodeError("offsetMinutes out of range", path)
    return _obj(
        [
            ("fold", str(value.fold)),
            ("offsetMinutes", str(offset)),
            ("tag", '"datetime"'),
            ("value", json_string(text)),
        ]
    )


def _encode_uuid(value: UUID, path: str) -> str:
    text = str(value)
    if not grammar.is_valid_uuid(text):
        raise CodecEncodeError("invalid uuid", path)
    return _obj([("tag", '"uuid"'), ("value", json_string(text))])


def _encode_path(value: PurePath, path: str) -> str:
    flavor = "windows" if isinstance(value, PureWindowsPath) else "posix"
    text = str(value)
    if not grammar.is_canonical_relative_path(text, flavor):
        raise CodecEncodeError("invalid or absolute path", path)
    return _obj(
        [
            ("flavor", json_string(flavor)),
            ("tag", '"path"'),
            ("value", json_string(text)),
        ]
    )


def _encode_sequence(tag: str, items: list[Any], budget: Budget, seen: set[int], path: str) -> str:
    with _container(items, seen, path):
        budget.enter()
        parts = [_encode_node(item, budget, seen, f"{path}.items[{i}]") for i, item in enumerate(items)]
        budget.leave()
    return _obj([("items", "[" + ",".join(parts) + "]"), ("tag", json_string(tag))])


def _encode_set(tag: str, items: Any, budget: Budget, seen: set[int], path: str) -> str:
    with _container(items, seen, path):
        budget.enter()
        encoded = [_encode_node(item, budget, seen, f"{path}.items[{i}]") for i, item in enumerate(items)]
        budget.leave()
    encoded.sort(key=lambda text: text.encode("utf-8"))
    for i in range(1, len(encoded)):
        if encoded[i - 1] == encoded[i]:
            raise CodecEncodeError(f"duplicate {tag} member", path)
    return _obj([("items", "[" + ",".join(encoded) + "]"), ("tag", json_string(tag))])


def _encode_dict(value: dict[Any, Any], budget: Budget, seen: set[int], path: str) -> str:
    with _container(value, seen, path):
        budget.enter()
        entries: list[tuple[bytes, str]] = []
        for i, (key, val) in enumerate(value.items()):
            key_text = _encode_node(key, budget, seen, f"{path}.entries[{i}].key")
            val_text = _encode_node(val, budget, seen, f"{path}.entries[{i}].value")
            entry = _obj([("key", key_text), ("value", val_text)])
            entries.append((key_text.encode("utf-8"), entry))
        budget.leave()
    entries.sort(key=lambda pair: pair[0])
    for i in range(1, len(entries)):
        if entries[i - 1][0] == entries[i][0]:
            raise CodecEncodeError("duplicate dict key", path)
    return _obj([("entries", "[" + ",".join(entry for _, entry in entries) + "]"), ("tag", '"dict"')])


def _encode_record(
    node: Any,
    record_type: str,
    name: str,
    names: list[str],
    values: list[Any],
    budget: Budget,
    seen: set[int],
    path: str,
) -> str:
    with _container(node, seen, path):
        budget.enter()
        seen_names: set[str] = set()
        fields: list[str] = []
        for i, (field_name, field_value) in enumerate(zip(names, values)):
            if field_name in seen_names:
                raise CodecEncodeError("duplicate record field name", path)
            seen_names.add(field_name)
            fields.append(
                _obj(
                    [
                        ("name", _encode_str(field_name, f"{path}.fields[{i}].name")),
                        ("value", _encode_node(field_value, budget, seen, f"{path}.fields[{i}].value")),
                    ]
                )
            )
        budget.leave()
    return _obj(
        [
            ("fields", "[" + ",".join(fields) + "]"),
            ("name", _encode_str(name, f"{path}.name")),
            ("tag", '"record"'),
            ("type", json_string(record_type)),
        ]
    )


def _encode_exception(value: BaseException, budget: Budget, seen: set[int], path: str) -> str:
    with _container(value, seen, path):
        details_value = getattr(value, "details", None)
        details = "null" if details_value is None else _encode_node(details_value, budget, seen, f"{path}.details")
    return _obj(
        [
            ("details", details),
            ("message", _encode_str(str(value), f"{path}.message")),
            ("tag", '"exception"'),
            ("type", _encode_str(type(value).__name__, f"{path}.type")),
        ]
    )


def _encode_list_node(value: ListNode, budget: Budget, seen: set[int], path: str) -> str:
    with _container(value, seen, path):
        values, cycle_index = list_to_canonical(value)
        budget.enter()
        encoded = [_encode_node(v, budget, seen, f"{path}.values[{i}]") for i, v in enumerate(values)]
        budget.leave()
    return _obj(
        [
            ("cycleIndex", "null" if cycle_index is None else str(cycle_index)),
            ("tag", '"ListNode"'),
            ("values", "[" + ",".join(encoded) + "]"),
        ]
    )


def _encode_tree_node(value: TreeNode, budget: Budget, seen: set[int], path: str) -> str:
    with _container(value, seen, path):
        slots = tree_to_canonical(value)
        budget.enter()
        parts = [
            "null" if slot is None else _encode_node(slot, budget, seen, f"{path}.values[{i}]")
            for i, slot in enumerate(slots)
        ]
        budget.leave()
    return _obj([("tag", '"TreeNode"'), ("values", "[" + ",".join(parts) + "]")])


class _container:
    """Context manager pushing a container onto the cycle-detection set."""

    __slots__ = ("_node_id", "_seen", "_path")

    def __init__(self, node: Any, seen: set[int], path: str) -> None:
        self._node_id = id(node)
        self._seen = seen
        self._path = path

    def __enter__(self) -> "_container":
        if self._node_id in self._seen:
            raise CodecEncodeError("reference cycle detected", self._path)
        self._seen.add(self._node_id)
        return self

    def __exit__(self, *exc: Any) -> None:
        self._seen.discard(self._node_id)
